[dtensor][5/n] switch random ops to op strategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109145
* #109144
* #108264
* __->__ #108263

This PR switches the random ops to use op strategy instead of rule
based, this is a first series of PRs to refactor ops after we refactor
op dispatch logic

wanchaol committed:
[dtensor][5/n] switch random ops to op strategy

This PR switches the random ops to use op strategy instead of rule
based, this is a first series of PRs to refactor ops after we refactor
op dispatch logic

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/n] switch random ops to op strategy"


This PR switches the random ops to use op strategy instead of rule
based, this is a first series of PRs to refactor ops after we refactor
op dispatch logic

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -170,8 +172,14 @@ def spec_to_strategy(spec: object) -> object:
                 suggestion_schema = [reshard_schema]
 
+            output_spec: OutputSpecType = output_strategy.output_spec
+            if op_schema.op == aten.native_dropout.default:
+                # for ops return multiple tensors, make output spec return same spec
+                # returned from the op strategy
+                output_spec = (output_strategy.output_spec, output_strategy.output_spec)
```
Can we have this in the random_op_strategy?

wanchaol replied:
Just tried, but it's a bit non-trival to make the `output_spec` to be a tuple of dtensor spec and resharding becomes complicated. I'll try to refactor this in a more canonical way

wanchaol committed:
Update on "[dtensor][5/n] switch random ops to op strategy"


This PR switches the random ops to use op strategy instead of rule
based, this is a first series of PRs to refactor ops after we refactor
op dispatch logic

[ghstack-poisoned]

fduwjj commented:
Also another n00b question, does this mean we want to move all rules, ideally, to strategies? Because keeping two ways of registering op seems to be a little bit confusing to users who want to register their own op?

wanchaol commented:
> Also another n00b question, does this mean we want to move all rules, ideally, to strategies? Because keeping two ways of registering op seems to be a little bit confusing to users who want to register their own op?

Yep! the north star is that all ops going forward can be implemented with strategy instead of rule, that would make our stack cleaner with possible easier op implementation

wanchaol committed:
Update on "[dtensor][5/n] switch random ops to op strategy"


This PR switches the random ops to use op strategy instead of rule
based, this is a first series of PRs to refactor ops after we refactor
op dispatch logic

[ghstack-poisoned]

fduwjj reviewed (approved):
Unblock for now, and looking forward to more refactoring.
